src/hammer-tech/stackup.py

```diff
@@ -0,0 +1,230 @@
+
+    def __str__(self) -> str:
+        return reverse_dict(RoutingDirection.__mapping())[self]
+
+    def opposite(self) -> "RoutingDirection":
+        if self == RoutingDirection.Vertical:
+            return RoutingDirection.Horizontal
+        elif self == RoutingDirection.Horizontal:
+            return RoutingDirection.Vertical
+        else:
+            return self
+
+# Note to maintainers: This is mirrored in schema.json- don't change one without the other
+
+# A tuple of wire width limit and spacing for generating a piecewise linear rule for spacing based on wire width
+# width_at_least: Any wires larger than this must obey the minSpacing rule
+# min_spacing: The minimum spacing for this bin. If a wire is wider than multiple entries, the worst-case (larger) minSpacing wins.
+class WidthSpacingTuple(NamedTuple('WidthSpacingTuple', [
+    ('width_at_least', float),
+    ('min_spacing', float)
+])):
+    __slots__ = ()
+
+    @staticmethod
+    def from_setting(d: dict) -> "WidthSpacingTuple":
+        w = float(d["width_at_least"])
+        s = float(d["min_spacing"])
+        assert w >= 0.0
+        assert s > 0.0
+        return WidthSpacingTuple(
+            width_at_least=w,
+            min_spacing=s
+        )
+
+    @staticmethod
+    def from_list(l: List[dict]) -> List["WidthSpacingTuple"]:
+        out = sorted(list(map(lambda x: WidthSpacingTuple.from_setting(x), l)), key=lambda x: x.width_at_least)
+        s = 0.0
+        for wst in out:
+            assert wst.min_spacing >= s
+            s = wst.min_spacing
+        return out
+
+
+# A metal layer and some basic info about it
+# name: M1, M2, etc.
+# index: The order in the stackup (lower is closer to the substrate)
+# direction: The preferred routing direction of this metal layer, or "redistribution" for non-routing top-level redistribution metals like Aluminum
+# min_width: The minimum wire width for this layer
+# pitch: The minimum cross-mask pitch for this layer (NOT same-mask pitch for multiply-patterned layers)
+# offset: The routing track offset from the origin for the first track in this layer (0 = first track is on an axis)
+# power_strap_widths_and_spacings: A list of WidthSpacingTuples that specify the minimum spacing rules for an infinitely long wire of variying width
+class Metal(NamedTuple('Metal', [
+    ('name', str),
+    ('index', int),
+    ('direction', "RoutingDirection"),
+    ('min_width', float),
+    ('pitch', float),
+    ('offset', float),
+    ('power_strap_widths_and_spacings', List[WidthSpacingTuple])
+])):
+    __slots__ = ()
+
+    # TODO this is assuming that a manufacturing grid is 0.001
+    def grid_unit(self) -> float:
+        return 0.001
+
+    # TODO internally represent numbers as integers or fixed-point to obviate the need for this
+    # ucb-bar/hammer#319
+    def snap(self, x: float) -> float:
+        return float(round(x / self.grid_unit())) * self.grid_unit()
```

Comment:
Do you need the float conversion? I would have assumed `int` * `float` = `float`

Reply:
Probably not.